File: backend/src/main/java/com/shopverse/service/StatsService.java
package com.shopverse.service;

import com.shopverse.repository.OrderRepository;
import com.shopverse.repository.ProductRepository;
import com.shopverse.repository.UserRepository;
import org.springframework.stereotype.Service;

import java.util.Map;

@Service
public class StatsService {

    private final OrderRepository orderRepository;
    private final ProductRepository productRepository;
    private final UserRepository userRepository;
    private final OrderService orderService;
    private final ProductService productService;

    public StatsService(OrderRepository orderRepository,
                        ProductRepository productRepository,
                        UserRepository userRepository,
                        OrderService orderService,
                        ProductService productService) {
        this.orderRepository = orderRepository;
        this.productRepository = productRepository;
        this.userRepository = userRepository;
        this.orderService = orderService;
        this.productService = productService;
    }

    public Map<String, Object> stats() {
        var lowStock = productRepository.findLowStock(5, org.springframework.data.domain.PageRequest.of(0, 6));
        return Map.of(
                "revenue", productRepository.sumRevenue(),
                "orders_count", orderRepository.count(),
                "customers_count", userRepository.count(),
                "products_count", productRepository.countByIsActive(true),
                "pending_orders", orderRepository.countByStatus(com.shopverse.domain.OrderStatus.PENDING),
                "low_stock_products", lowStock.stream().map(productService::toDto).toList(),
                "low_stock_count", lowStock.size(),
                "recent_orders", orderService.allOrders().stream().limit(5).toList(),
                "orders_by_status", Map.ofEntries(
                        Map.entry("PENDING", orderRepository.countByStatus(com.shopverse.domain.OrderStatus.PENDING)),
                        Map.entry("PAID", orderRepository.countByStatus(com.shopverse.domain.OrderStatus.PAID)),
                        Map.entry("SHIPPED", orderRepository.countByStatus(com.shopverse.domain.OrderStatus.SHIPPED)),
                        Map.entry("DELIVERED", orderRepository.countByStatus(com.shopverse.domain.OrderStatus.DELIVERED)),
                        Map.entry("CANCELLED", orderRepository.countByStatus(com.shopverse.domain.OrderStatus.CANCELLED))
                )
        );
    }
}

File: backend/src/services/statsService.js
'use strict'

const db = require('../db')
const productService = require('./productService')
const orderService = require('./orderService')

async function stats() {
  const revenue = await db.get(
    `SELECT COALESCE(SUM(total_amount), 0) AS revenue FROM orders
     WHERE status IN ('PAID', 'SHIPPED', 'DELIVERED')`,
  )
  const ordersCount = await db.get('SELECT COUNT(*) AS cnt FROM orders')
  const customersCount = await db.get(
    `SELECT COUNT(*) AS cnt FROM users WHERE role = 'ROLE_CUSTOMER'`,
  )
  const productsCount = await db.get(
    `SELECT COUNT(*) AS cnt FROM products WHERE is_active = ${db.boolLit(true)}`,
  )
  const pendingOrders = await db.get(`SELECT COUNT(*) AS cnt FROM orders WHERE status = 'PENDING'`)
  const lowStock = await productService.findLowStock(5, 50)

  const byStatusRows = await db.q('SELECT status, COUNT(*) AS cnt FROM orders GROUP BY status')
  const ordersByStatus = { PENDING: 0, PAID: 0, SHIPPED: 0, DELIVERED: 0, CANCELLED: 0 }
  for (const row of byStatusRows) {
    if (row.status in ordersByStatus) ordersByStatus[row.status] = Number(row.cnt)
  }

  const recent = await orderService.recentOrders(5)

  return {
    revenue: Number(revenue.revenue) || 0,
    orders_count: Number(ordersCount.cnt),
    customers_count: Number(customersCount.cnt),
    products_count: Number(productsCount.cnt),
    pending_orders: Number(pendingOrders.cnt),
    low_stock_products: lowStock.map(productService.toDto),
    low_stock_count: lowStock.length,
    recent_orders: recent,
    orders_by_status: ordersByStatus,
  }
}

module.exports = { stats }
